### PR TITLE
syntax/parse: configurable failed-match error

### PR DESCRIPTION
**Pull request by mflatt**
CC @rmculpepper 

Add `syntax/parse/report-config` with `current-report-configuration` to support configuring parts of the message that `syntax-parse` generates for a failed match. Configuration controls the way that datum and identifier literals are described and printed.

The existing `~describe` form can generally configure the way that expected input is described in an error message, but for the common cases of literals (including keywords), an external configuration can be more convenient and seems likely more efficient. This configuration is enough to make Rhombus error messages better, covering the key cases of keywords, identifiers, and literals. It doesn't cover operator matching for Rhombus, since operators have the shape `(op <identifier>)` in S-expression form; using `~describe` for the specific case of operators seems fine. Although Rhombus obviously motivates this change, the goal here is to be reasonably general and to leave room when further generalization is needed.

This PR also includes a use of `error-syntax->string-handler` instead of `~s` to format a source term.

### CODE REVIEW

**Comment by mflatt** [comment]
Would need a version bump and `history` note before merging.

File: racket/collects/syntax/parse/private/runtime-report.rkt [hunk 1]
```diff
@@ -620,61 +622,64 @@ This suggests the following new algorithm based on (s):
...
+                (report (prose-for-expects (expect:disj-expects frame-expect) config)
                         context frame-stx within-stx)]
                [else
-                (report (prose-for-expects (list frame-expect))
+                (report (prose-for-expects (list frame-expect) config)
                         context frame-stx within-stx)])]))
 
-;; prose-for-expects : (Listof Expect) -> String
-(define (prose-for-expects expects)
+;; prose-for-expects : (Listof Expect) Hash-Table -> String
```
[the remark this answers is not in the document]
Comment (mflatt) [reply racket/collects/syntax/parse/private/runtime-report.rkt]: fixed

File: racket/collects/syntax/parse/private/runtime-report.rkt [hunk 2]
```diff
@@ -687,21 +692,26 @@ This suggests the following new algorithm based on (s):
...
 ;; prose-for-first-desc : FirstDesc -> String
-(define (prose-for-first-desc desc)
+(define ((prose-for-first-desc config) desc)
   (match desc
     [(? string?) desc]
     [(list 'any) "any term"] ;; FIXME: maybe should cancel out other descs ???
-    [(list 'literal id) (format "the identifier `~s'" id)]
-    [(list 'datum (? symbol? s)) (format "the literal symbol `~s'" s)]
-    [(list 'datum d) (format "the literal ~s" d)]))
+    [(list 'datum id) (format "the ~a ~a"
```
[the remark this answers is not in the document]
Comment (mflatt) [reply racket/collects/syntax/parse/private/runtime-report.rkt]: fixed

File: racket/collects/syntax/parse/private/runtime-report.rkt [hunk 3]
```diff
@@ -711,13 +721,24 @@ This suggests the following new algorithm based on (s):
...
-                               #:max-width 50))
+                           ((error-syntax->string-handler) stx 50))
                        (~a " location: "
                            (or (source-location->string stx) "not available")))
                  null)))]))
 
+(define (singular what/s) (if (pair? what/s) (car what/s) what/s))
+(define (plural what/s) (if (pair? what/s) (cadr what/s) what/s))
+
+(define (parition-by-what vs whats)
```
[the remark this answers is not in the document]
Comment (mflatt) [reply racket/collects/syntax/parse/private/runtime-report.rkt]: Changed to use `group-by`

**Comment by mflatt** [comment]
Thanks for the corrections!

File: pkgs/racket-doc/syntax/scribblings/parse/error.scrbl [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+  The default configuration formats a symbol value using
+  @racket[(format "`~s'" v)], and it formats an identifier the same
+  after extracting its symbol with @racket[syntax-e].}
+
+ ]
+
+}
+
+
+@defproc[(report-configuration? [v any/c]) boolean?]{
```
Comment (rmculpepper) [inline pkgs/racket-doc/syntax/scribblings/parse/error.scrbl:76]: It doesn't look like this is exported.

**Comment by rmculpepper** [comment]
One more issue, but other than that feel free to merge when you're ready.

File: pkgs/racket-doc/syntax/scribblings/parse/error.scrbl (hunk 4 above)
Comment (mflatt) [reply pkgs/racket-doc/syntax/scribblings/parse/error.scrbl:76]: fixed (with test)

**Pull request merged**